### ISSUE TITLE
k3s kubectl logs pod failed behind proxy

### ISSUE THREAD

**Issue by vincentmli** [k3s-io/k3s#5712]
**Environmental Info:**
K3s Version: 

Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.1+k3s1", GitCommit:"0581808f5c160b0c0cafec5b8f20430835f34f44", GitTreeState:"clean", BuildDate:"2022-06-11T17:26:28Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.1+k3s1", GitCommit:"0581808f5c160b0c0cafec5b8f20430835f34f44", GitTreeState:"clean", BuildDate:"2022-06-11T17:26:28Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"linux/amd64"}

Node(s) CPU architecture, OS, and Version: 

Cluster Configuration:

**Describe the bug:**

when running k3s behind proxy, kubectl fail to get pod log

cat /root/.bash_profile 
```
#esnet proxy
PROXY_URL="http://10.3.254.254:3128/"
export http_proxy="$PROXY_URL"
export https_proxy="$PROXY_URL"
export no_proxy="127.0.0.1,10.0.0.0/8,10.3.0.0/16,10.169.72.0/24,localhost"
```

install k3s

```
root@cilium-demo-1:/home#  curl -sfL https://get.k3s.io | INSTALL_K3S_SYMLINK=force INSTALL_K3S_VERSION='v1.24.1+k3s1' INSTALL_K3S_EXEC='--disable=traefik --disable-network-policy' sh -

[INFO]  Using v1.24.1+k3s1 as release
[INFO]  Downloading hash https://github.com/k3s-io/k3s/releases/download/v1.24.1+k3s1/sha256sum-amd64.txt
[INFO]  Downloading binary https://github.com/k3s-io/k3s/releases/download/v1.24.1+k3s1/k3s
[INFO]  Verifying binary download
[INFO]  Installing k3s to /usr/local/bin/k3s
[INFO]  Skipping installation of SELinux RPM
[INFO]  Creating /usr/local/bin/kubectl symlink to k3s
[INFO]  Creating /usr/local/bin/crictl symlink to k3s
[INFO]  Creating /usr/local/bin/ctr symlink to k3s
[INFO]  Creating killall script /usr/local/bin/k3s-killall.sh
[INFO]  Creating uninstall script /usr/local/bin/k3s-uninstall.sh
[INFO]  env: Creating environment file /etc/systemd/system/k3s.service.env
[INFO]  systemd: Creating service file /etc/systemd/system/k3s.service
[INFO]  systemd: Enabling k3s unit
Created symlink /etc/systemd/system/multi-user.target.wants/k3s.service → /etc/systemd/system/k3s.service.
[INFO]  systemd: Starting k3s

```
pod running

```
root@cilium-demo-1:/home# kubectl get po -o wide -A
NAMESPACE     NAME                                      READY   STATUS    RESTARTS   AGE   IP          NODE            NOMINATED NODE   READINESS GATES
kube-system   local-path-provisioner-7b7dc8d6f5-4g26s   1/1     Running   0          88s   10.42.0.3   cilium-demo-1   <none>           <none>
kube-system   coredns-b96499967-m9qsk                   1/1     Running   0          88s   10.42.0.2   cilium-demo-1   <none>           <none>
kube-system   metrics-server-668d979685-sqfrz           1/1     Running   0          88s   10.42.0.4   cilium-demo-1   <none>           <none>

```
logs pod

```
root@cilium-demo-1:/home# kubectl logs coredns-b96499967-m9qsk  -n kube-system

Error from server: Get "https://cilium-demo-1:10250/containerLogs/kube-system/coredns-b96499967-m9qsk/coredns": Service Unavailable
```

**Steps To Reproduce:**

**Expected behavior:**
able to log pod behind proxy

**Actual behavior:**
unable to log pod behind proxy

**Additional context / logs:**

**Backporting**

- [ ] Needs backporting to older releases

**Comment by brandond** [comment]
Can you reinstall with `INSTALL_K3S_EXEC='--disable=traefik --disable-network-policy --debug', re-run your `kubectl logs` command, and attach the k3s service logs, as well as the contents of `/var/lib/rancher/k3s/agent/containerd/containerd.log`?

**Comment by vincentmli** [comment]
[k3s.service.log.txt](https://github.com/k3s-io/k3s/files/8904676/k3s.service.log.txt)


Jun 15 00:06:48 cilium-demo-1 k3s[1495]: time="2022-06-15T00:06:48Z" level=debug msg="Tunnel server egress proxy dialing 10.3.254.254:3128 directly"

Jun 15 00:06:48 cilium-demo-1 k3s[1495]: E0615 00:06:48.341914    1495 status.go:71] apiserver received an error that is not an metav1.Status: &url.Error{Op:"Get", URL:"https://cilium-demo-1:10250/containerLogs/kube-system/coredns-b96499967-8gv2c/coredns", Err:(*errors.errorString)(0xc007f877d0)}: Get "https://cilium-demo-1:10250/containerLogs/kube-system/coredns-b96499967-8gv2c/coredns": Service Unavailable

no containerd.log

```
 ls -l /var/lib/rancher/k3s/agent/containerd/
total 52
drwx--x--x 2 root root 4096 Jun 15 00:05 bin
drwxr-xr-x 4 root root 4096 Jun 15 00:05 io.containerd.content.v1.content
drwxr-xr-x 4 root root 4096 Jun 15 00:06 io.containerd.grpc.v1.cri
drwx--x--x 2 root root 4096 Jun 15 00:05 io.containerd.metadata.v1.bolt
drwx--x--x 2 root root 4096 Jun 15 00:05 io.containerd.runtime.v1.linux
drwx--x--x 3 root root 4096 Jun 15 00:05 io.containerd.runtime.v2.task
drwx------ 2 root root 4096 Jun 15 00:05 io.containerd.snapshotter.v1.btrfs
drwx------ 3 root root 4096 Jun 15 00:05 io.containerd.snapshotter.v1.fuse-overlayfs
drwx------ 3 root root 4096 Jun 15 00:05 io.containerd.snapshotter.v1.native
drwx------ 3 root root 4096 Jun 15 00:05 io.containerd.snapshotter.v1.overlayfs
drwx------ 4 root root 4096 Jun 15 00:05 io.containerd.snapshotter.v1.stargz
drwx--x--x 2 root root 4096 Jun 15 00:05 lib
drwx------ 2 root root 4096 Jun 15 00:06 tmpmounts
```

**Comment by brandond** [comment]
Hmm, so it's trying to connect to the kubelet via the configured HTTP proxy, by way of the apiserver egress. That is interesting.

**Comment by vincentmli** [comment]
it works with your suggestion with flannel

Jun 15 02:18:13 cilium-demo-1 k3s[1544]: time="2022-06-15T02:18:13Z" level=debug msg="Tunnel server handing HTTP/1.1 CONNECT request for //10.3.72.9:10250 from 127.0.0.1:45014"

Jun 15 02:18:13 cilium-demo-1 k3s[1544]: time="2022-06-15T02:18:13Z" level=debug msg="Tunnel server egress proxy dialing 127.0.0.1:10250 directly"

it also work with cilium cni now

`curl -sfL https://get.k3s.io | INSTALL_K3S_SYMLINK=force INSTALL_K3S_VERSION='v1.24.1+k3s1' INSTALL_K3S_EXEC='--flannel-backend=none --node-ip=10.169.72.9 --disable=traefik --disable-network-policy --kube-apiserver-arg=kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname --debug' sh -`

`cilium install --kube-proxy-replacement strict --helm-set-string="k8sServiceHost=10.169.72.9,k8sServicePort=6443"`

logs

```
kubectl logs cilium-2m9q9  -n kube-system | head -1
Defaulted container "cilium-agent" out of: cilium-agent, mount-cgroup (init), clean-cilium-state (init)
level=info msg="Started gops server" address="127.0.0.1:9890" subsys=daemon
```

**Comment by vincentmli** [comment]
sorry one more question, I tried to add agent node, but the agent node always got node ip with the ip `10.3.72.19` that has default route setup on the VM even if I specified `--node-ip=10.169.72.19`

agent node

`
curl -sfL https://get.k3s.io | K3S_URL='https://10.169.72.9:6443' INSTALL_K3S_SYMLINK=force INSTALL_K3S_VERSION='v1.24.1+k3s1' INSTALL_K3S_EXEC='--flannel-backend=none --node-ip=10.169.72.19 --disable=traefik --disable-network-policy --kube-apiserver-arg=kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname' K3S_TOKEN="K105950e3b839dd35501bcaab259925cad51186195861b6ebd822e5ad96f7151a4d::server:7c91ee12b4a4bd9029121f95bce51bb6" sh -`

two node:

```
 kubectl get no -o wide
NAME            STATUS     ROLES                  AGE   VERSION        INTERNAL-IP   EXTERNAL-IP   OS-IMAGE           KERNEL-VERSION      CONTAINER-RUNTIME
cilium-demo-3   NotReady   <none>                 11m   v1.24.1+k3s1   10.3.72.19    <none>        Ubuntu 22.04 LTS   5.15.0-37-generic   containerd://1.6.6-k3s1
cilium-demo-1   Ready      control-plane,master   24m   v1.24.1+k3s1   10.169.72.9   <none>        Ubuntu 22.04 LTS   5.15.0-37-generic   containerd://1.6.6-k3s1
```
I prefer agent node `cilium-demo-3` on node ip `10.169.72.19`, is that possible? this behind proxy thing just make it too complicate :)

**Comment by vincentmli** [comment]
I need to add `--node-external-ip` to the control node

`
 curl -sfL https://get.k3s.io | INSTALL_K3S_SYMLINK=force INSTALL_K3S_VERSION='v1.24.1+k3s1' INSTALL_K3S_EXEC='--flannel-backend=none --node-ip=10.169.72.9 --node-external-ip=10.3.72.9 --disable=traefik --disable-network-policy --kube-apiserver-arg=kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname' sh -
`

then agent node

`
curl -sfL https://get.k3s.io | K3S_URL='https://10.3.72.9:6443' INSTALL_K3S_SYMLINK=force INSTALL_K3S_VERSION='v1.24.1+k3s1' INSTALL_K3S_EXEC='--node-ip=10.169.72.19 --node-external-ip=10.3.72.19' K3S_TOKEN="K1047760bfa51d578138b55562fbb51a3b737a28be2f8f0269a58971ae8775e6723::server:f551e7cfa93f07d2947dd2e0e18ee2a7" sh -
`

node

```
 kubectl get node -o wide
NAME            STATUS   ROLES                  AGE     VERSION        INTERNAL-IP    EXTERNAL-IP   OS-IMAGE           KERNEL-VERSION      CONTAINER-RUNTIME
cilium-demo-1   Ready    control-plane,master   12m     v1.24.1+k3s1   10.169.72.9    10.3.72.9     Ubuntu 22.04 LTS   5.15.0-37-generic   containerd://1.6.6-k3s1
cilium-demo-3   Ready    <none>                 3m49s   v1.24.1+k3s1   10.169.72.19   10.3.72.19    Ubuntu 22.04 LTS   5.15.0-37-generic   containerd://1.6.6-k3s1

```

cilium agent

```
root@cilium-demo-1:~# kubectl get po -o wide -A
NAMESPACE     NAME                                      READY   STATUS    RESTARTS   AGE     IP             NODE            NOMINATED NODE   READINESS GATES
kube-system   cilium-operator-757f586845-f5x5k          1/1     Running   0          12m     10.169.72.9    cilium-demo-1   <none>           <none>
kube-system   cilium-zbzjc                              1/1     Running   0          12m     10.169.72.9    cilium-demo-1   <none>           <none>
kube-system   local-path-provisioner-7b7dc8d6f5-5wm5b   1/1     Running   0          12m     10.0.0.150     cilium-demo-1   <none>           <none>
kube-system   coredns-b96499967-mczk6                   1/1     Running   0          12m     10.0.0.173     cilium-demo-1   <none>           <none>
kube-system   metrics-server-668d979685-mldzk           1/1     Running   0          12m     10.0.0.60      cilium-demo-1   <none>           <none>
kube-system   cilium-8wbdb                              1/1     Running   0          4m34s   10.169.72.19   cilium-demo-3   <none>           <none>
```

agent node cilium agent log

```
kubectl logs  cilium-8wbdb  -n kube-system | head -1
Defaulted container "cilium-agent" out of: cilium-agent, mount-cgroup (init), clean-cilium-state (init)
level=info msg="Started gops server" address="127.0.0.1:9890" subsys=daemon
```

**Comment by brandond** [comment]
This behavior is a little unexpected; I'll have to do some checking to see if there's any way to improve the default.

**Comment by ShylajaDevadiga** [comment]
Validated on v1.24.2-rc1+k3s2
## Setup 1
## Environment Details 
*Infrastructure*
[ *] Cloud

*Node(s) CPU architecture, OS, and Version:*

Ubuntu 20:.04

 
*Cluster Configuration:*

Single node

## Replication Steps
**Reproduced on**
k3s v1.24.1+k3s1
 ```
curl -fL https://get.k3s.io | INSTALL_K3S_VERSION=v1.24.1+k3s1 sh -s - server
```
Assign k3s env variables as below
```
sudo cat /etc/systemd/system/k3s.service.env
HTTPS_PROXY="http://proxy.example.com"
HTTP_PROXY="http://proxy.example.com"
NO_PROXY="172.31.0.0/16,127.0.0.1,localhost"
```
Restart k3s
```
sudo systemctl restart k3s
```
Access the logs 
```
$ kubectl logs -n kube-system coredns-b96499967-zlm2q  
Error from server: Get "https://ip-172-31-14-170:10250/containerLogs/kube-system/coredns-b96499967-zlm2q/coredns": Not Found
```

## Validation Steps
 **Validated on**
k3s v1.24.2-rc1+k3s2
```
curl -fL https://get.k3s.io | INSTALL_K3S_VERSION=v1.24.2-rc1+k3s2 sh -s - server
```
Assign k3s env variables as below
```
sudo cat /etc/systemd/system/k3s.service.env
HTTPS_PROXY="http://proxy.example.com"
HTTP_PROXY="http://proxy.example.com"
NO_PROXY="172.31.0.0/16,127.0.0.1,localhost"
```
Restart k3s
```
sudo systemctl restart k3s
```
Access the logs 
```
 $ kubectl logs -n kube-system coredns-b96499967-nl9pd    
[WARNING] No files matching import glob pattern: /etc/coredns/custom/*.server
.:53
[WARNING] No files matching import glob pattern: /etc/coredns/custom/*.server
[INFO] plugin/reload: Running configuration SHA512 = b941b080e5322f6519009bb49349462c7ddb6317425b0f6a83e5451175b720703949e3f3b454a24e77f3ffe57fd5e9c6130e528a5a1dd00d9000e4afd6c1108d
CoreDNS-1.9.1
```

## Setup 2: Passing bind-address

## Environment Details 
*Infrastructure*
[ *] Cloud

*Node(s) CPU architecture, OS, and Version:*
Ubuntu 20:.04

*Cluster Configuration:*
Single node


## Replication Steps
**Reproduced on**
k3s v1.24.1+k3s1

```
$ cat /etc/rancher/k3s/config.yaml 
token: test
write-kubeconfig-mode: 644
bind-address: <node-ip>
```

 ```
curl -fL https://get.k3s.io | INSTALL_K3S_VERSION=v1.24.1+k3s1 sh -s - server
```
Assign k3s env variables as below
```
sudo cat /etc/systemd/system/k3s.service.env
HTTPS_PROXY="http://proxy.example.com"
HTTP_PROXY="http://proxy.example.com"
NO_PROXY="172.31.0.0/16,127.0.0.1,localhost"
```
Restart k3s
```
sudo systemctl restart k3s
```
Access the logs
```
$ kubectl logs -n kube-system coredns-b96499967-bjj4b
Error from server: Get "https://ip-172-31-1-204:10250/containerLogs/kube-system/coredns-b96499967-bjj4b/coredns": proxyconnect tcp: dial tcp 127.0.0.1:6443: connect: connection refused
```

## Validation Steps
 **Validated on**
k3s v1.24.2-rc1+k3s2
```
$ cat /etc/rancher/k3s/config.yaml 
token: test
write-kubeconfig-mode: 644
bind-address: <node-ip>
```
```
curl -fL https://get.k3s.io | INSTALL_K3S_VERSION=v1.24.2-rc1+k3s2 sh -s - server
```
Assign k3s env variables as below
```
sudo cat /etc/systemd/system/k3s.service.env
HTTPS_PROXY="http://proxy.example.com"
HTTP_PROXY="http://proxy.example.com"
NO_PROXY="172.31.0.0/16,127.0.0.1,localhost"
```
Restart k3s
```
sudo systemctl restart k3s
```
Access the logs 
```
 $ kubectl logs -n kube-system coredns-b96499967-5tc45     
[WARNING] No files matching import glob pattern: /etc/coredns/custom/*.server
[INFO] plugin/ready: Still waiting on: "kubernetes"
.:53
[WARNING] No files matching import glob pattern: /etc/coredns/custom/*.server
[INFO] plugin/reload: Running configuration SHA512 = b941b080e5322f6519009bb49349462c7ddb6317425b0f6a83e5451175b720703949e3f3b454a24e77f3ffe57fd5e9c6130e528a5a1dd00d9000e4afd6c1108d
CoreDNS-1.9.1

```

**Comment by dlouzan** [comment]
@brandond We are using a corporate proxy and just upgraded two single-node clusters from `1.23.6+k3s1` to `1.24.1+k3s1` and started to experience the issue, but contrary to the release notes, we are not able to solve it by upgrading to `1.24.2+k3s2`.

Actually tried a couple of different versions, including `1.23.8+k3s2` which also mentions solving the problem on 1.23.x, to no avail. The only solution at this moment is to revert to `1.23.6+k3s1`.

```sh
# cat /etc/systemd/system/k3s.service.env
http_proxy="<corporate-proxy>"
https_proxy="http://<corporate-proxy>"
no_proxy="localhost,127.0.0.1,<some-internal-domains>"

# kubectl logs deploy/some-deploy
Error from server: Get "https://<cluster-dns>:10250/containerLogs/<pod>": proxyconnect tcp: proxy error from 127.0.0.1:6443 while dialing <corporate-proxy>, code 500: 500 Internal Server Error
```

Any ideas? :bow:

**Comment by brandond** [comment]
If you don't want k3s to go through the proxy to access your nodes, ensure that your node address space is covered by the no_proxy environment variable: `no_proxy="localhost,127.0.0.1,<some-internal-domains>,<some-internal-cidrs>"`

For example: `no_proxy="localhost,127.0.0.1,corp.example.com,10.0.0.0/8"`

**Comment by dlouzan** [comment]
@brandond Thanks for the answer! I had actually tried that in my tests but did not solve it either, though I can give it another try tomorrow.

E.g. we deploy with (ansible syntax):

```sh
get-k3s.sh --data-dir {{ system_k3s_dir }} --default-local-storage-path {{ system_pv_dir }} --disable=traefik --cluster-cidr "172.16.0.0/16" --service-cidr "172.17.0.0/16" --cluster-dns "172.17.0.10"
```

and then added to the proxy settings on `k3s.service.env`:

```sh
no_proxy="localhost,127.0.0.1,<some-internal-domains>,172.16.0.0/16,172.17.0.0/16"
```

Then restarted the k3s server, but did not make a difference.

**Comment by brandond** [comment]
The cluster and service CIDRs are already added to the no_proxy variable automatically, I was referring to the actual CIDR blocks for the LAN segments that the nodes are on.

**Comment by max-wittig** [comment]
I think adding internal cidr ranges should not be required, as it also was not required before.

This is a breaking change in a minor release ☹️

**Comment by dlouzan** [comment]
@brandond Thanks again for the hints, yup, adding the internal CIDRs to `no_proxy` did solve the issue. I still don't understand why is that needed now if that wasn't in the past, I share the same concerns as @max-wittig about a breaking change on a minor release.

Also a couple of extra points:

- Via https://github.com/k3s-io/k3s/pull/5603, my understanding for the new flag ` --egress-selector-mode` is that previously this was defaulted to `agent` on k3s, and now it is `pod`. So for the sake of fully understanding the issue, I tried reverting the flag to `agent`, but interestingly this did not solve the problem either, the `no_proxy` extra setting for the internal CIDRs is needed in any case.
- Additionally, shouldn't this new flag be documented in https://rancher.com/docs/k3s/latest/en/installation/install-options/agent-config/?

**Comment by brandond** [comment]
Yeah, that's fair. Previous releases of K3s carried a patch to core Kubernetes that replaced the dialer that the apiserver used to access the kubelet, with a replacement that had proxy support disabled and used remotedialer. 

Now that we've dropped that patch (and we're always trying to drop patches), we're seeing the same behavior as upstream vanilla Kubernetes - which is that the apiserver will try to use the proxy to access the kubelet, if one is configured. I'll have to see if there's a better way to handle this, without bring back some or all of that patch.

**Comment by vincentmli** [comment]
@brandond thanks for making k3s a great project, I use k3s daily as test lab for my OSS cilium contribution and the company I work for also use k3s :-)

**Comment by dlouzan** [comment]
@brandond Just created https://github.com/k3s-io/docs/pull/7 to add the documentation :bow:

**Comment by svscorp** [comment]
@brandond 

In version 1.30.2+k3s2, I am still having this issue

The control node sits at 10.20.20.1 (and configured to be bound to 10.20.20.1)
1) added no_proxy="localhost,127.0.0.1,10.20.20.0/24,172.16.0.0/16,172.17.0.0/16" into the `/etc/systemd/system/k3s.service.env` file
2) reloaded daemon, restarted - still
Error from server: Get "https://10.20.20.1:10250/containerLogs/kube-system/...": proxy error from 10.20.20.1:6443 while dialing 10.20.20.1:10250, code 502: 502 Bad Gateway

Also, joining the message thanks for making k3s a great project :)

**Comment by brandond** [comment]
@svscorp this is a 2 year old resolved issue, whatever is going on with your node is not the same. Please open a new issue, and fill out the issue template.
